File: src/cubyz/world/terrain/biomes/Boulder.java
package cubyz.world.terrain.biomes;

import cubyz.api.Resource;
import cubyz.utils.FastRandom;
import cubyz.world.Chunk;
import cubyz.world.blocks.Blocks;
import cubyz.world.terrain.CaveMap;
import pixelguys.json.JsonObject;

/**
 * Generates stone boulders of various sizes.
 */
public class Boulder extends StructureModel {
	private final int block;
	private final float size, sizeVariation;

	public Boulder() {
		super(new Resource("cubyz", "boulder"), 0);
		block = 0;
		size = 0;
		sizeVariation = 0;
	}

	public Boulder(JsonObject json) {
		super(new Resource("cubyz", "boulder"), json.getFloat("chance", 0.5f));
		block = Blocks.getByID(json.getString("block", "cubyz:stone"));
		size = json.getFloat("size", 4);
		sizeVariation = json.getFloat("size_variation", 1);
	}

	@Override
	public void generate(int x, int z, int y, Chunk chunk, CaveMap map, FastRandom rand) {
		float radius = size + sizeVariation*(rand.nextFloat()*2 - 1);
		// My basic idea is to use a point cloud and a potential function to achieve somewhat smooth boulders without being a sphere.
		final int numberOfPoints = 4;
		float[] pointCloud = new float[numberOfPoints*3];
		for(int i = 0; i < pointCloud.length; i++) {
			pointCloud[i] = (rand.nextFloat() - 0.5f)*radius/2;
		}
		// My potential functions is ¹⁄ₙ Σ (radius/2)²/(x⃗ - x⃗ₚₒᵢₙₜ)²
		// This ensures that the entire boulder is inside of a square with sidelength 2*radius.
		for(int px = chunk.startIndex((int)(x - radius)); px <= (int)(x + radius); px += chunk.voxelSize) {
			for(int py = chunk.startIndex((int)(y - radius)); py <= (int)(y + radius); py += chunk.voxelSize) {
				for(int pz = chunk.startIndex((int)(z - radius)); pz <= (int)(z + radius); pz += chunk.voxelSize) {
					if (!chunk.liesInChunk(px, py, pz)) continue;
					float potential = 0;
					for(int i = 0; i < numberOfPoints; i++) {
						float dx = px - x - pointCloud[3*i];
						float dy = py - y - pointCloud[3*i + 1];
						float dz = pz - z - pointCloud[3*i + 2];
						float distSqr = dx*dx + dy*dy + dz*dz;
						potential += 1/distSqr;
					}
					potential *= radius*radius/4/numberOfPoints;
					if(potential >= 1) {
							chunk.updateBlockInGeneration(px, py, pz, block);
					}
				}
			}
		}
	}

	@Override
	public StructureModel loadStructureModel(JsonObject json) {
		return new Boulder(json);
	}
}

File: src/server/terrain/simple_structures/Boulder.zig
const std = @import("std");

const main = @import("root");
const random = main.random;
const ZonElement = main.ZonElement;
const terrain = main.server.terrain;
const CaveMap = terrain.CaveMap;
const vec = main.vec;
const Vec3d = vec.Vec3d;
const Vec3f = vec.Vec3f;
const Vec3i = vec.Vec3i;
const NeverFailingAllocator = main.utils.NeverFailingAllocator;

pub const id = "cubyz:boulder";

pub const generationMode = .floor;

const Boulder = @This();

blockType: u16,
size: f32,
sizeVariation: f32,

pub fn loadModel(arenaAllocator: NeverFailingAllocator, parameters: ZonElement) *Boulder {
	const self = arenaAllocator.create(Boulder);
	self.* = .{
		.blockType = main.blocks.getByID(parameters.get([]const u8, "block", "cubyz:stone")),
		.size = parameters.get(f32, "size", 4),
		.sizeVariation = parameters.get(f32, "size_variation", 1),
	};
	return self;
}

pub fn generate(self: *Boulder, x: i32, y: i32, z: i32, chunk: *main.chunk.ServerChunk, caveMap: terrain.CaveMap.CaveMapView, seed: *u64, _: bool) void {
	_ = caveMap;
	const radius = self.size + self.sizeVariation*(random.nextFloat(seed)*2 - 1);
	// My basic idea is to use a point cloud and a potential function to achieve somewhat smooth boulders without being a sphere.
	const numberOfPoints = 4;
	var pointCloud: [numberOfPoints]Vec3f = undefined;
	for(&pointCloud) |*point| {
		point.* = Vec3f {
			(random.nextFloat(seed) - 0.5)*radius/2,
			(random.nextFloat(seed) - 0.5)*radius/2,
			(random.nextFloat(seed) - 0.5)*radius/2,
		};
	}
	// My potential functions is ¹⁄ₙ Σ (radius/2)²/(x⃗ - x⃗ₚₒᵢₙₜ)²
	// This ensures that the entire boulder is inside of a square with sidelength 2*radius.
	const maxRadius: i32 = @intFromFloat(@ceil(radius));
	var px = chunk.startIndex(x - maxRadius);
	while(px < x + maxRadius) : (px += chunk.super.pos.voxelSize) {
		var py = chunk.startIndex(y - maxRadius);
		while(py < y + maxRadius) : (py += chunk.super.pos.voxelSize) {
			var pz = chunk.startIndex(z - maxRadius);
			while(pz < z + maxRadius) : (pz += chunk.super.pos.voxelSize) {
				if(!chunk.liesInChunk(px, py, pz)) continue;
				var potential: f32 = 0;
				for(&pointCloud) |point| {
					const delta = @as(Vec3f, @floatFromInt(Vec3i{px, py, pz} - Vec3i{x, y, z})) - point;
					const distSqr = vec.dot(delta, delta);
					potential += 1/distSqr;
				}
				potential *= radius*radius/4/numberOfPoints;
				if(potential >= 1) {
					chunk.updateBlockInGeneration(px, py, pz, .{.typ = self.blockType, .data = 0}); // TODO: Natural standard.
				}
			}
		}
	}
}
